{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from IPython.core.display_functions import display\n",
    "\n",
    "table_dict = {\n",
    "    'accessionID' : [],\n",
    "    'gene' : [],\n",
    "    'variant': [],\n",
    "    'RNA' : [],\n",
    "    'comments' : []\n",
    "}\n",
    "\n",
    "gene_idx = -1\n",
    "rna_idx = -1\n",
    "variant_num = -1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "for record in SeqIO.parse(\"transcriptome.fasta\", \"fasta\"):\n",
    "    parsed_record = re.split(', |; ', record.description)\n",
    "\n",
    "    table_dict['accessionID'].append(record.id)\n",
    "\n",
    "    # Logic implemented to handle cases where additional commas are added in the gene description or alternative\n",
    "    # variant names, etc.\n",
    "    # Ex: NR_133665.2 Homo sapiens mucin 17, cell surface associated (MUC17), transcript variant 2, also known as V0, non-coding RNA\n",
    "    # It has been verified that gene_idx and rna_idx are always redefined for this dataset\n",
    "    for i in range(len(parsed_record)):\n",
    "        if parsed_record[i][-1] == \")\":\n",
    "            gene_idx = i\n",
    "        if gene_idx != -1:\n",
    "            if parsed_record[i][-3:] == 'RNA':\n",
    "                rna_idx = i\n",
    "            if parsed_record[i][-1].isnumeric():\n",
    "                # Not all variant's are single numbers, but they do end with one Ex: transcript variant X3\n",
    "                # Only recording variant names, not alternative names\n",
    "                variant_num = parsed_record[i].split()[-1]\n",
    "\n",
    "    # Finding the gene abbrev.\n",
    "    table_dict['gene'].append(parsed_record[gene_idx][parsed_record[gene_idx].find(\"(\")+1:parsed_record[gene_idx].find(\")\")])\n",
    "\n",
    "    table_dict['RNA'].append(parsed_record[rna_idx])\n",
    "\n",
    "    # Checking if we defined variant_num, as records can have no variant defined\n",
    "    table_dict['variant'].append(variant_num if variant_num != -1 else None)\n",
    "\n",
    "    # Checking if there is parsed data after the RNA, which indicates a comment, verified that if condition is true there is a comment\n",
    "    table_dict['comments'].append(parsed_record[rna_idx+1] if rna_idx + 1 != len(parsed_record) else None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accessionID</th>\n",
       "      <th>gene</th>\n",
       "      <th>variant</th>\n",
       "      <th>RNA</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>A2M</td>\n",
       "      <td>1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000015.3</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000016.6</td>\n",
       "      <td>ACADM</td>\n",
       "      <td>1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>nuclear gene for mitochondrial product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000017.4</td>\n",
       "      <td>ACADS</td>\n",
       "      <td>1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>nuclear gene for mitochondrial product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000018.4</td>\n",
       "      <td>ACADVL</td>\n",
       "      <td>1</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>nuclear gene for mitochondrial product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555358</th>\n",
       "      <td>XR_953251.3</td>\n",
       "      <td>LOC105378536</td>\n",
       "      <td>X1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555359</th>\n",
       "      <td>XR_953252.3</td>\n",
       "      <td>LOC105378536</td>\n",
       "      <td>X4</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555360</th>\n",
       "      <td>XR_953253.3</td>\n",
       "      <td>LOC105378536</td>\n",
       "      <td>X5</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555361</th>\n",
       "      <td>XR_953307.2</td>\n",
       "      <td>LOC105370752</td>\n",
       "      <td>X2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555362</th>\n",
       "      <td>XR_953308.2</td>\n",
       "      <td>LOC105370752</td>\n",
       "      <td>X5</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>555363 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        accessionID          gene variant    RNA  \\\n",
       "0       NM_000014.6           A2M       1   mRNA   \n",
       "1       NM_000015.3          NAT2       1   mRNA   \n",
       "2       NM_000016.6         ACADM       1   mRNA   \n",
       "3       NM_000017.4         ACADS       1   mRNA   \n",
       "4       NM_000018.4        ACADVL       1   mRNA   \n",
       "...             ...           ...     ...    ...   \n",
       "555358  XR_953251.3  LOC105378536      X1  ncRNA   \n",
       "555359  XR_953252.3  LOC105378536      X4  ncRNA   \n",
       "555360  XR_953253.3  LOC105378536      X5  ncRNA   \n",
       "555361  XR_953307.2  LOC105370752      X2  ncRNA   \n",
       "555362  XR_953308.2  LOC105370752      X5  ncRNA   \n",
       "\n",
       "                                      comments  \n",
       "0                                         None  \n",
       "1                                         None  \n",
       "2       nuclear gene for mitochondrial product  \n",
       "3       nuclear gene for mitochondrial product  \n",
       "4       nuclear gene for mitochondrial product  \n",
       "...                                        ...  \n",
       "555358                                    None  \n",
       "555359                                    None  \n",
       "555360                                    None  \n",
       "555361                                    None  \n",
       "555362                                    None  \n",
       "\n",
       "[555363 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transcriptome_df = pd.DataFrame(data=table_dict)\n",
    "display(transcriptome_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
